File: convalida-validators/src/main/java/convalida/validators/AbstractValidator.java
package convalida.validators;

import android.widget.EditText;

import convalida.validators.util.EditTexts;

import static convalida.validators.util.EditTexts.removeError;
import static convalida.validators.util.EditTexts.setError;

/**
 * @author Wellington Costa on 21/06/2017.
 */
public abstract class AbstractValidator {

    public final EditText editText;
    public final String errorMessage;
    public final boolean autoDismiss;
    private boolean hasError = false;

    public AbstractValidator(EditText editText, String errorMessage, boolean autoDismiss) {
        this.editText = editText;
        this.errorMessage = errorMessage;
        this.autoDismiss = autoDismiss;
    }

    protected abstract boolean isValid(String value);

    private void executeValidation(String value) {
        hasError = !EditTexts.isVisible(editText) && !isValid(value);

        if (hasError) {
            setError(editText, errorMessage);
        } else {
            removeError(editText);
        }
    }

    public boolean validate() {
        executeValidation(editText.getText().toString());
        return !hasError;
    }

    public boolean validate(String value) {
        executeValidation(value);
        return !hasError;
    }

    public void clear() {
        removeError(editText);
    }


}

File: convalida-validators/src/main/java/convalida/validators/AbstractValidator.kt
package convalida.validators

import android.view.View
import android.widget.EditText
import convalida.validators.util.EditTextUtils

/**
 * @author Wellington Costa on 21/06/2017.
 */
abstract class AbstractValidator(private val editText: EditText, private val errorMessage: String, autoDismiss: Boolean) : Validator {

    private var hasError: Boolean = false

    internal abstract fun isNotValid(value: String): Boolean

    private fun executeValidation(value: String) {
        hasError = if(viewIsVisible()) isNotValid(value) else false

        if (hasError) {
            EditTextUtils.setError(editText, errorMessage)
        } else {
            EditTextUtils.setError(editText, null)
        }
    }

    private fun viewIsVisible(): Boolean {
        return !(editText.visibility == View.GONE ||
                editText.visibility == View.INVISIBLE)
    }

    override fun validate(): Boolean {
        executeValidation(editText.text.toString())
        return !hasError
    }

    override fun clear() {
        EditTextUtils.setError(editText, null)
    }

    init {
        this.hasError = false
        if(autoDismiss) {
            EditTextUtils.addOnTextChangedListener(editText, ::executeValidation)
        }
    }

}